examples/tidalfarm/tidalfarm.py

```diff
@@ -1,119 +1,189 @@
+options.t_end = 0.5*tidal_period
+options.outputdir = 'outputs'
+options.nonlin = False
+options.include_grad_depth_viscosity_term = False
 options.check_vol_conservation_2d = True
-options.fields_to_export = ['uv_2d', 'elev_2d']
 solver_obj.options.timestepper_type = 'cranknicolson'
-solver_obj.options.shallow_water_theta = 1.0
+solver_obj.options.shallow_water_theta = 0.6
+options.h_viscosity = Constant(60.)
+options.quadratic_drag = Constant(0.0025)
 solver_obj.options.solver_parameters_sw = {
     'mat_type': 'aij',
     'ksp_type': 'preonly',
     'pc_type': 'lu',
     'pc_factor_mat_solver_package': 'mumps',
-    'snes_monitor': False,
+    'snes_monitor': True,
     'snes_type': 'newtonls',
 }
-options.dt = timestep  # override computed dt
-options.h_viscosity = Constant(2.0)
-
-# create function spaces
-solver_obj.create_function_spaces()
 
-# create drag function and set something there
-drag_func = Function(solver_obj.function_spaces.P1_2d, name='bottomdrag')
-x = SpatialCoordinate(mesh2d)
-drag_center = 12.0
-drag_bg = 0.0025
-x0 = lx/2
-y0 = ly/2
-sigma = 20.0
-drag_func.project(drag_center*exp(-((x[0]-x0)**2 + (x[1]-y0)**2)/sigma**2) + drag_bg, annotate=False)
-# assign fiction field
-options.quadratic_drag = drag_func
-
-velocity_u = 2.0
 # assign boundary conditions
-inflow_tag = 1
-outflow_tag = 2
-inflow_bc = {'un': Constant(-velocity_u)}  # NOTE negative into domain
-outflow_bc = {'elev': Constant(0.0)}
+left_tag = 1
+right_tag = 2
+coasts_tag = 3
+tidal_elev = Function(FunctionSpace(mesh2d, "CG", 1), name='tidal_elev')
+tidal_elev_bc = {'elev': tidal_elev}
+noslip_bc = {'uv': Constant((0.0, 0.0))}
+solver_obj.bnd_functions['shallow_water'] = {
+    left_tag: tidal_elev_bc,
+    right_tag: tidal_elev_bc,
+    coasts_tag: noslip_bc
+}
 
-solver_obj.bnd_functions['shallow_water'] = {inflow_tag: inflow_bc,
-                                             outflow_tag: outflow_bc}
 
-solver_obj.assign_initial_conditions(uv=as_vector((velocity_u, 0.0)))
-solver_obj.iterate()
+# first setup all the usual SWE terms
+solver_obj.create_equations()
 
-adj_html("forward.html", "forward")
-adj_html("adjoint.html", "adjoint")
+# defines an additional turbine drag term to the SWE
+turbine_friction = Function(FunctionSpace(mesh2d, "CG", 1), name='turbine_friction')
 
 
-integral = solver_obj.fields.solution_2d[0]*dx
-J = Functional(integral*dt[FINISH_TIME], name="MyFunctional")
-c = Control(drag_func)
-dJdc = compute_gradient(J, c, forget=False)
-out = File('gradient_J.pvd')
-out.write(dJdc)
-J0 = assemble(integral)
-print_output("Functional evaluated by hand: ", J0)
+class TurbineDragTerm(shallowwater_eq.ShallowWaterMomentumTerm):
+    def residual(self, uv, eta, uv_old, eta_old, fields, fields_old, bnd_conditions=None):
+        total_h = self.get_total_depth(eta_old)
+        C_D = turbine_friction
+        f = C_D * sqrt(dot(uv_old, uv_old)) * inner(self.u_test, uv) / total_h * self.dx(2)
+        return -f
```

Comment:
Yes, it would be good to discuss this. There are various bits in this example that I would like to move out (eventually) and offer as generic functionality. For the tidal drag here it's indeed just a standard quadratic drag term but applied over a subdomain (the farm areas). There will be various extensions of this however:
* we often do this for multiple farm areas, so you'd have a multiple of these terms
* in the discrete approach, where we optimise for the position of individual turbines, the turbine friction is not a function but an expression that basically creates individual localised bump functions at the position of each turbine
* for this discrete approach, there is a correction term that approximately translates between the depth-averaged model velocity and either the actual velocity through the turbine or the theoretical free stream velocity. This correction depends on the depth
* in the more realistic case where the thrust coefficient is not constant (thrust curves), the drag is not really quadratic any more as the friction coefficient itself also depends on the local velocity

So in general this could all get a little more messy, so I was thinking of having this sort of stuff in a separate module - but we could also have a generic turbine drag term in the SWE module where we allow the coefficient to be any sort of expression possibly dependent on uv_old and depth, and allow it to either be integrated over the entire domain (for the discrete approach) or over a sum of subdomains. I'm not sure that would be cleaner though.